File: Java_Virtual_World/src/Gameplay/Animals/Animal.java
package Gameplay.Animals;

import Gameplay.Organism;
import java.lang.Math;
import java.util.Vector;

public abstract class Animal extends Organism {
    public Animal(int x, int y, int s, int i){
        super(x, y, s, i);
    }
    private int prevX;
    private int prevY;
    private boolean didMultiply;

    @Override
    public void Action(){
        randomMove(1);
    }
    @Override
    public void Collision(){
        Organism tmp = world.getCollisionOrg(this);
        if(tmp != null){
            if(tmp.toString().equals(toString())){
                multiply((Animal) tmp);
            }
            else{
                fight(tmp);
            }
        }
    }
    @Override
    public void NextRound(){
        didMultiply = false;
    }
    protected void randomMove(int range){
        if(isSmelling() && areEveryoneStronger()) { return; }

        int[] cords = {-1 * range, 0, range};
        int randomX = cords[(int) (Math.random()*3)];
        int randomY = cords[(int) (Math.random()*3)];
        while ((randomX + x == x && randomY + y == y)
            || !world.isInsideMap(x + randomX, y + randomY)
            || (isSmelling() && world.isOrganismAt(x + randomX, y + randomY) != null && world.isOrganismAt(x + randomX, y + randomY).getStrength() > strength)
        ){
             randomX = cords[(int) (Math.random() * 3)];
             randomY = cords[(int) (Math.random() * 3)];
        }
            changeLocation(x+randomX, y+randomY);
            return;
    }
    protected void changeLocation(int x, int y){
        if(world.isInsideMap(x,y)) {
            prevX = this.x;
            prevY = this.y;
            this.x = x;
            this.y = y;
        }
    }
    private boolean areEveryoneStronger() {
        if(world.isOrganismAt(x-1, y) != null){
            Organism tmp = world.isOrganismAt(x-1, y);
            if(tmp.getStrength() <= strength){ return false; }
        }
        if(world.isOrganismAt(x+1, y) != null){
            Organism tmp = world.isOrganismAt(x+1, y);
            if(tmp.getStrength() <= strength){ return false; }
        }
        if(world.isOrganismAt(x, y - 1) != null){
            Organism tmp = world.isOrganismAt(x, y - 1);
            if(tmp.getStrength() <= strength){ return false; }
        }
        if(world.isOrganismAt(x, y + 1) != null){
            Organism tmp = world.isOrganismAt(x, y + 1);
            if(tmp.getStrength() <= strength){ return false; }
        }

        return true;
    }
    private void reverse(){
        x = prevX;
        y = prevY;
    }
    private void multiply(Animal crush) {
        if(crush.getAge() == 0){ return; }

        Organism child = Copy();
        reverse();
        Vector<Integer> cords = world.getFreeSquare(crush.getX(), crush.getY());
        if((cords.get(0) == crush.getX() && cords.get(1) == crush.getY()) || didMultiply || crush.didMultiply){ return; }

        child.setX(cords.get(0));
        child.setY(cords.get(1));
        world.addOrg(child);
        didMultiply = true;
        crush.didMultiply = true;
    }
    private void fight(Organism opponent){
        if(escape() || opponent.escape()){ return; }
        if(opponent.didBlocked(opponent)){ reverse(); return; }
        if(strength < opponent.getStrength()){
            kill();
            world.getNots().addNode(opponent.toString() + "has killed" + toString() + "at (" + x + "," + y + ")" );
        }
        else{
            opponent.kill();
            world.getNots().addNode(toString() + "has killed" + opponent.toString() + "at (" + x + "," + y + ")" );
        }
    }
}

File: Python_Virtual_World/Animal.py
from ast import BitAnd
import copy
from random import randint
from MyVector import MyVector
from Organism import Organism


class Animal(Organism):


    def __init__(self, position : MyVector, s : int, i : int):

        super().__init__(position, s, i)
        self._didMultiply = False
        self._prev = position


    def action(self):

        self._RandomMove()


    def collision(self):

        opponent = self._world.getCollision(self)

        if opponent is None:
            return

        if str(opponent) == str(self):

            self._multiply(opponent)
            return

        self._fight(opponent)


    def newRound(self):

        self._didMultiply = False



    def _RandomMove(self, range = 1):


        if self.isSmelling() and self.everyoneStronger():
            return

        cords = [-1 * range, 0,range]
        prev = MyVector(self._position.getY(), self._position.getX())

        while True:

            randX = cords[randint(0,2)]
            randY = cords[randint(0,2)]

            d = MyVector(randY,randX)

            self._changeLocation(d)

            if not (prev == self._position or
                    (self.isSmelling() and
                     self._world.getCollision(self) is not None and
                     self._world.getCollision(self).getStrength() > self.getStrength())):
                break

    def _changeLocation(self, d : MyVector):

        if not (self.getPosition() + d) \
            .outside(self._world.getHeight(), self._world.getWidth()):

            self._prev = MyVector(self._position.getY(), self._position.getX())
            self._position += d


    def _multiply(self, crush):

        if crush.getAge() == 0:
            return

        org = copy.deepcopy(self)
        self.__reverse()

        birthField = self._world.getFreeSquare(crush.getPosition())

        if birthField == crush.getPosition() or crush._didMultiply or self._didMultiply:
            return


        org.setPosition(birthField)
        org.setAge(-1)

        self._world.addOrg(org)
        self._world.getNots().add(org.__str__() + " was born!")

        self._didMultiply = True
        crush._didMultiply = True


    def __reverse(self):
        self.setPosition(self._prev)

    def _fight(self, opponent : Organism):

        if self.escape() or opponent.escape():
            return

        if self.getStrength() < opponent.getStrength():

            if self.didBlocked(opponent):

                self.__reverse()
                return

            self._world.getNots().add(f"{str(opponent)} has eaten {str(self)}")
            self.kill()
            self.giveMod(opponent)

            return

        if opponent.didBlocked(self):

            self.__reverse()
            return

        self._world.getNots().add(f"{str(self)} has eaten {str(opponent)}")
        opponent.kill()
        opponent.giveMod(self)

    def everyoneStronger(self) -> bool:

        for y in range(-1,2):
            for x in range(-1,2):

                field = MyVector(y,x)

                org = self._world.getOrganismAt(self.getPosition() + field)

                if org != self and (org is None or org.getStrength() <= self.getStrength()):

                    return False


        return True




